byzcoin/api.go

```diff
@@ -85,6 +87,70 @@ func (c *Client) getLatestKnownBlock() *skipchain.SkipBlock {
 	return c.Latest
 }
 
+type nodeBlock struct {
+	si *network.ServerIdentity
+	sb *skipchain.SkipBlock
+}
+
+// UpdateNodes asks all nodes for their last skipblock and creates a list of the
+// nodes with the highest block number.
+func (c *Client) UpdateNodes() {
+	nodes := make(chan nodeBlock, 1)
+	scCl := skipchain.NewClient()
+	latestID := c.ID
+	if c.Latest != nil {
+		latestID = c.Latest.SkipChainID()
+	}
+
+	// Get the latest block from all nodes.
+	for _, si := range c.Roster.List {
+		go func(si *network.ServerIdentity) {
+			r := onet.NewRoster([]*network.ServerIdentity{si})
+			reply, err := scCl.GetUpdateChain(r, latestID)
+			if err != nil {
+				log.Lvl2("Error in GetUpdateChain:", err)
+				return
+			}
+			nodes <- nodeBlock{si, reply.Update[len(reply.Update)-1]}
+		}(si)
+	}
+
+	// Wait for all nodes to reply or for the timeout to happen
+	var nodeList []nodeBlock
+	for i := 0; i < len(c.Roster.List); i++ {
+		select {
+		case n := <-nodes:
+			if c.Latest == nil || n.sb.Index > c.Latest.Index {
+				c.Latest = n.sb
+				c.Roster = *n.sb.Roster
+			}
+			nodeList = append(nodeList, n)
+		case <-time.After(time.Second * 2):
```

Comment:
The loop will timeout iff there is no new `nodeBlock` for two seconds, so it might takes as long as `2s * len(c.Roster.List)`; wanted?

Reply:
I'm never sure what the `break` does here - won't it break from the `for`-loop?

Reply:
After discussion, yes, if the answers trickle in, that is OK.